### PR TITLE
plic: Recode to use OH knowledge

### PR DESCRIPTION
**Pull request by mwachs5**
This reduces the lines of code in the PLIC by 30%. Not yet tested.

### CODE REVIEW

File: src/main/scala/uncore/devices/Plic.scala [hunk 1]
```diff
@@ -182,14 +176,28 @@ class TLPLIC(params: PLICParams)(implicit p: Parameters) extends LazyModule
       PLICConsts.enableBase(i) -> e.map(b => RegField(1, b))
     }
 
+    val claimer = Wire(init = Vec.fill(nHarts){Bool(false)})
+    val claiming = Wire(init = Vec.tabulate(nHarts){i => Mux(claimer(i), UIntToOH(maxDevs(i)), UInt(0, width=log2Up(nDevices+1)))})
+    val claimedDevs = Wire(init = Vec(claiming.reduceLeft( _ | _ ).toBools))
+
+    for ((pg, c) <- (pending zip gateways) zip claimedDevs) {
```
Comment (aswaterman) [inline src/main/scala/uncore/devices/Plic.scala]: ```for (((p, g), c) <- (pending zip gateways) zip claimedDevs) {```

then you don't need the ```val p``` and ```val g```

File: src/main/scala/uncore/devices/Plic.scala [hunk 2]
```diff
@@ -182,14 +176,28 @@ class TLPLIC(params: PLICParams)(implicit p: Parameters) extends LazyModule
...
+      val p = pg._1
+      val g = pg._2
+      g.ready := !p
+      g.complete := false
+      when(c) {
+        p := false
+      }.elsewhen (g.valid) {
+        p := true
+      }
```
Comment (aswaterman) [inline src/main/scala/uncore/devices/Plic.scala]: Given that the goal is to reduce line count, it's worth seeing if this does better:

    when (c || g.valid) { p := !c }

Comment (mwachs5) [inline src/main/scala/uncore/devices/Plic.scala]: I tried it out, overall LoC is not affected, they just get turned from if-else into assigns.

Comment (mwachs5) [inline src/main/scala/uncore/devices/Plic.scala]: Actually I was wrong, it does help the LoC considerably.

**Comment by mwachs5** [comment]
This passes tests.

File: src/main/scala/uncore/devices/Plic.scala [hunk 3]
```diff
@@ -182,27 +176,60 @@ class TLPLIC(params: PLICParams)(implicit p: Parameters) extends LazyModule
       PLICConsts.enableBase(i) -> e.map(b => RegField(1, b))
     }
 
+    // When a hart reads a claim/complete register, then the
+    // device which is currently its highest priority is no longer pending.
+    // This code expolits the fact that, practically, only one claim/complete
```
Comment (colinschmidt) [inline src/main/scala/uncore/devices/Plic.scala]: expolits -> exploits

**Pull request merged**